### PR TITLE
BasicUI: fix possible NPE in widget renderer, fixes #2379

### LINKED ISSUE #2379: BASIC UI throws webserver error 500 when parsing NULL value
When I'm getting a NULL value from one of my sensors, the BASIC UI kills the webserver:

```
02:29:42.482 [WARN ] [eclipse.jetty.servlet.ServletHandler] - /basicui/app
java.lang.NullPointerException
        at java.net.URLEncoder.encode(URLEncoder.java:204)[:1.8.0_111]
        at org.eclipse.smarthome.ui.basic.internal.render.AbstractWidgetRenderer.escapeURL(AbstractWidgetRenderer.java:166)[184:org.eclipse.smarthome.ui.basic:0.9.0.201610252044]
        at org.eclipse.smarthome.ui.basic.internal.render.AbstractWidgetRenderer.getState(AbstractWidgetRenderer.java:212)[184:org.eclipse.smarthome.ui.basic:0.9.0.201610252044]
        at org.eclipse.smarthome.ui.basic.internal.render.AbstractWidgetRenderer.preprocessSnippet(AbstractWidgetRenderer.java:86)[184:org.eclipse.smarthome.ui.basic:0.9.0.201610252044]
        at org.eclipse.smarthome.ui.basic.internal.render.TextRenderer.renderWidget(TextRenderer.java:43)[184:org.eclipse.smarthome.ui.basic:0.9.0.201610252044]
        at org.eclipse.smarthome.ui.basic.internal.render.PageRenderer.renderWidget(PageRenderer.java:166)[184:org.eclipse.smarthome.ui.basic:0.9.0.201610252044]
        at org.eclipse.smarthome.ui.basic.internal.render.PageRenderer.processChildren(PageRenderer.java:128)[184:org.eclipse.smarthome.ui.basic:0.9.0.201610252044]
        at org.eclipse.smarthome.ui.basic.internal.render.PageRenderer.processChildren(PageRenderer.java:149)[184:org.eclipse.smarthome.ui.basic:0.9.0.201610252044]
        at org.eclipse.smarthome.ui.basic.internal.render.PageRenderer.processPage(PageRenderer.java:92)[184:org.eclipse.smarthome.ui.basic:0.9.0.201610252044]
        at org.eclipse.smarthome.ui.basic.internal.servlet.WebAppServlet.service(WebAppServlet.java:150)[184:org.eclipse.smarthome.ui.basic:0.9.0.201610252044]
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:812)[81:org.eclipse.jetty.servlet:9.2.14.v20151106]
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:587)[81:org.eclipse.jetty.servlet:9.2.14.v20151106]
        at org.ops4j.pax.web.service.jetty.internal.HttpServiceServletHandler.doHandle(HttpServiceServletHandler.java:70)[171:org.ops4j.pax.web.pax-web-jetty:4.2.4]
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)[80:org.eclipse.jetty.server:9.2.14.v20151106]
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:577)[79:org.eclipse.jetty.security:9.2.14.v20151106]
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:223)[80:org.eclipse.jetty.server:9.2.14.v20151106]
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)[80:org.eclipse.jetty.server:9.2.14.v20151106]
        at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext.doHandle(HttpServiceContext.java:271)[171:org.ops4j.pax.web.pax-web-jetty:4.2.4]
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)[81:org.eclipse.jetty.servlet:9.2.14.v20151106]
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)[80:org.eclipse.jetty.server:9.2.14.v20151106]
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)[80:org.eclipse.jetty.server:9.2.14.v20151106]
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)[80:org.eclipse.jetty.server:9.2.14.v20151106]
        at org.ops4j.pax.web.service.jetty.internal.JettyServerHandlerCollection.handle(JettyServerHandlerCollection.java:80)[171:org.ops4j.pax.web.pax-web-jetty:4.2.4]
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)[80:org.eclipse.jetty.server:9.2.14.v20151106]
        at org.eclipse.jetty.server.Server.handle(Server.java:499)[80:org.eclipse.jetty.server:9.2.14.v20151106]
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)[80:org.eclipse.jetty.server:9.2.14.v20151106]
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)[80:org.eclipse.jetty.server:9.2.14.v20151106]
        at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)[72:org.eclipse.jetty.io:9.2.14.v20151106]
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)[83:org.eclipse.jetty.util:9.2.14.v20151106]
        at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)[83:org.eclipse.jetty.util:9.2.14.v20151106]
        at java.lang.Thread.run(Thread.java:745)[:1.8.0_111]
```
Doing the same with the CLASSIC UI only gives a "NULL" at the item as expected.

### PR DESCRIPTION
**Pull request by resetnow**
fixes #2379

Signed-off-by: Vlad Ivanov <vlad-mbx@ya.ru>

### CODE REVIEW

File: extensions/ui/org.eclipse.smarthome.ui.basic/src/main/java/org/eclipse/smarthome/ui/basic/internal/render/AbstractWidgetRenderer.java [hunk 1]
```diff
@@ -202,14 +208,13 @@ protected String processColor(Widget w, String snippet) {
     }
 
     protected String getCategory(Widget w) {
-        String category = escapeURL(itemUIRegistry.getCategory(w));
```
Comment (kaikreuzer) [inline extensions/ui/org.eclipse.smarthome.ui.basic/src/main/java/org/eclipse/smarthome/ui/basic/internal/render/AbstractWidgetRenderer.java]: why do you remove the escaping here? How is that related to the bug report?

File: extensions/ui/org.eclipse.smarthome.ui.basic/src/main/java/org/eclipse/smarthome/ui/basic/internal/render/AbstractWidgetRenderer.java [hunk 2]
```diff
@@ -202,14 +208,13 @@ protected String processColor(Widget w, String snippet) {
...
-        return category;
+        return itemUIRegistry.getCategory(w);
     }
 
     protected String getState(Widget w) {
         State state = itemUIRegistry.getState(w);
         if (state != null) {
-            return escapeURL(state.toString());
```
Comment (kaikreuzer) [inline extensions/ui/org.eclipse.smarthome.ui.basic/src/main/java/org/eclipse/smarthome/ui/basic/internal/render/AbstractWidgetRenderer.java]: Why did you move the escaping to the preprocessSnippets method? Are there other callers of getState where you do not want escaping being done?

File: extensions/ui/org.eclipse.smarthome.ui.basic/src/main/java/org/eclipse/smarthome/ui/basic/internal/render/AbstractWidgetRenderer.java (hunk 1 above)
Comment (resetnow) [inline extensions/ui/org.eclipse.smarthome.ui.basic/src/main/java/org/eclipse/smarthome/ui/basic/internal/render/AbstractWidgetRenderer.java]: You're right about this one. The idea was that category is used in a html data attribute so it doesn't need URL escaping, but now I see it doesn't hurt to always escape it.

File: extensions/ui/org.eclipse.smarthome.ui.basic/src/main/java/org/eclipse/smarthome/ui/basic/internal/render/AbstractWidgetRenderer.java (hunk 2 above)
Comment (resetnow) [inline extensions/ui/org.eclipse.smarthome.ui.basic/src/main/java/org/eclipse/smarthome/ui/basic/internal/render/AbstractWidgetRenderer.java]: There are no other callers but I think semantically it makes sense to return the state as is, because there's nothing about this method name suggesting it returns a state escaped to use within an URL.

Comment (resetnow) [inline extensions/ui/org.eclipse.smarthome.ui.basic/src/main/java/org/eclipse/smarthome/ui/basic/internal/render/AbstractWidgetRenderer.java]: Same goes for getCategory. If it's ok with you, I'll add escaping to preprocessSnippet then. If not, I'll just revert this PR to add basic null checking for escapeURL method.

File: extensions/ui/org.eclipse.smarthome.ui.basic/src/main/java/org/eclipse/smarthome/ui/basic/internal/render/AbstractWidgetRenderer.java (hunk 1 above)
Comment (kaikreuzer) [inline extensions/ui/org.eclipse.smarthome.ui.basic/src/main/java/org/eclipse/smarthome/ui/basic/internal/render/AbstractWidgetRenderer.java]: afair, the category is used in the url to access the icon - so escaping might be relevant here.

File: extensions/ui/org.eclipse.smarthome.ui.basic/src/main/java/org/eclipse/smarthome/ui/basic/internal/render/AbstractWidgetRenderer.java (hunk 2 above)
Comment (kaikreuzer) [inline extensions/ui/org.eclipse.smarthome.ui.basic/src/main/java/org/eclipse/smarthome/ui/basic/internal/render/AbstractWidgetRenderer.java]: yep, ok to move it to the places, where escaping is really required.

**Comment by resetnow** [comment]
Updated.
